# Z-Tests Practice Workbook

Covers:
- One-sample Z-test (mean)
- One-sample proportion Z-test
- Two-proportion Z-test 

Each question follows: Hypothesis → Approach → Test Statistic → Decision → CI → Conclusion

In [1]:
import numpy as np
from scipy.stats import norm

---
## 📌 Question 1 — One-Sample Z-Test (Mean)

A factory claims that the mean diameter of its bolts is **10 mm**.

A sample of **50 bolts** is taken, and the following statistics are observed:

| Parameter | Value |
|-----------|-------|
| Sample mean (x̄) | 10.3 mm |
| Population std dev (σ) | 1.0 mm |
| Sample size (n) | 50 |
| Significance level (α) | 0.05 |

Perform a **two-tailed one-sample Z-test** to evaluate the factory's claim.
Also compute the **95% confidence interval** for the mean.

---

#### Step 1 — Hypothesis

- **H₀:** μ = 10 (factory's claim holds)
- **H₁:** μ ≠ 10 (two-tailed — we don't know which direction the mean might shift)

We're not assuming the bolt is too big or too small specifically, so two-tailed makes sense here.

#### Step 2 — Why Z-Test?

- Population σ is **known** → Z-test (if σ were unknown, we'd use t-test)
- n = 50, which is large enough for CLT to kick in
- We'll check both critical value and p-value methods to cross-verify

In [2]:
# Q1 given values
xbar_q1  = 10.3   # sample mean (mm)
mu0_q1   = 10     # hypothesized population mean
sig_q1   = 1      # known population std dev
n_q1     = 50     # sample size
a_q1     = 0.05   # significance level

#### Step 3 — Test Statistic

$$z = \frac{\bar{x} - \mu_0}{\sigma / \sqrt{n}}$$

We're measuring how many standard errors away the sample mean is from the claimed population mean.

In [3]:
z_q1 = (xbar_q1 - mu0_q1) / (sig_q1 / n_q1**0.5)
z_q1

2.1213203435596477

#### Step 4 — Critical Value Method

For α = 0.05 (two-tailed), we split the rejection region equally on both tails:

- Each tail gets α/2 = 0.025
- We need z such that P(Z < z) = 0.975 → use `norm.ppf(0.975)`

If |z_stat| > z_critical → reject H₀

In [4]:
# critical z at alpha=0.05, two-tailed
zcrit_q1 = norm.ppf(1 - a_q1/2)
zcrit_q1

np.float64(1.959963984540054)

**Decision:** |z| = 2.12 > z_critical = 1.96 → **Reject H₀**

#### Step 5 — p-value Method

$$p = 2 \times P(Z > |z|)$$

`norm.sf(x)` gives P(Z > x) directly — the survival function. Multiply by 2 for two-tailed.

In [5]:
# p-value for two-tailed test
pval_q1 = 2 * norm.sf(abs(z_q1))
pval_q1

np.float64(0.033894853524688816)

**Decision:** p-value ≈ 0.034 < α = 0.05 → **Reject H₀** ✓ (consistent with critical value method)

#### Step 6 — 95% Confidence Interval

$$CI = \bar{x} \pm z_{\alpha/2} \cdot \frac{\sigma}{\sqrt{n}}$$

If μ₀ falls outside this interval, that's another way to reject H₀.

In [6]:
moe_q1   = zcrit_q1 * (sig_q1 / np.sqrt(n_q1))  # margin of error
lo_q1    = xbar_q1 - moe_q1
hi_q1    = xbar_q1 + moe_q1
(lo_q1, hi_q1)

(np.float64(10.022819235130065), np.float64(10.577180764869937))

#### ✅ Final Conclusion — Q1

| Check | Result |
|-------|--------|
| z-stat | 2.12 |
| Critical value (±) | 1.96 |
| p-value | 0.034 |
| 95% CI | (10.02, 10.58) |

Both methods agree — the hypothesized mean of 10 mm lies **outside** the confidence interval, |z| exceeds the critical threshold, and p-value < α.

**We reject H₀.** There's sufficient statistical evidence that the true mean bolt diameter is **not 10 mm** — the factory's claim is questionable at the 5% significance level.

---
## 📌 Question 2 — One-Sample Proportion Z-Test

A survey of **800 voters** finds that **480 support** a new policy.
A pollster claims that **55% of the population** supports the policy.

| Parameter | Value |
|-----------|-------|
| Sample size (n) | 800 |
| Successes (x) | 480 |
| Hypothesized proportion (p₀) | 0.55 |
| Significance level (α) | 0.05 |

Perform a **two-tailed one-sample proportion Z-test**. Also compute the **95% CI**.

---

#### Step 1 — Hypothesis

- **H₀:** p = 0.55
- **H₁:** p ≠ 0.55 (two-tailed — the observed proportion could be higher or lower)

We're not trying to show the support is specifically greater or less, just whether it differs.

#### Step 2 — Why Proportion Z-Test?

- Data is count-based (successes out of n) → proportion test
- n is large → CLT applies, so we can use normal approximation
- Standard error under H₀ uses p₀ (not p̂), because we assume H₀ is true when testing

In [7]:
# Q2 given values
x2   = 480    # number of supporters
n2   = 800    # total surveyed
p0_2 = 0.55   # claimed population proportion
a2   = 0.05   # significance level

In [8]:
phat_2 = x2 / n2   # observed sample proportion
phat_2

0.6

#### Step 3 — Test Statistic

$$z = \frac{\hat{p} - p_0}{\sqrt{\frac{p_0(1-p_0)}{n}}}$$

SE uses p₀ here (not p̂) because under H₀ we assume the true proportion is p₀.

In [9]:
se_h0_2 = (p0_2 * (1 - p0_2) / n2) ** 0.5   # SE under null hypothesis
z_q2    = (phat_2 - p0_2) / se_h0_2
z_q2

2.8426762180748018

#### Step 4 — Critical Value Method

In [10]:
zcrit_q2 = norm.ppf(1 - a2/2)
zcrit_q2

np.float64(1.959963984540054)

**Decision:** 2.84 > 1.96 → |z| > z_critical → **Reject H₀**

#### Step 5 — p-value Method

In [11]:
pval_q2 = 2 * norm.sf(abs(z_q2))
pval_q2

np.float64(0.004473649256611822)

**Decision:** 0.0044 < 0.05 → p-value < α → **Reject H₀** ✓

#### Step 6 — 95% Confidence Interval

$$CI = \hat{p} \pm z_{\alpha/2} \cdot \sqrt{\frac{\hat{p}(1-\hat{p})}{n}}$$

**Note:** For CI we switch to p̂ in the SE formula — we're now estimating the true proportion, not assuming H₀.

In [12]:
se_est_2 = (phat_2 * (1 - phat_2) / n2) ** 0.5   # SE using p̂ for CI
moe_q2   = zcrit_q2 * se_est_2
lo_q2    = phat_2 - moe_q2
hi_q2    = phat_2 + moe_q2
(lo_q2, hi_q2)

(np.float64(0.5660524279777148), np.float64(0.6339475720222851))

#### ✅ Final Conclusion — Q2

| Check | Result |
|-------|--------|
| Sample proportion (p̂) | 0.60 |
| z-stat | 2.84 |
| Critical value (±) | 1.96 |
| p-value | 0.0044 |
| 95% CI | (0.566, 0.634) |

The claimed proportion p₀ = 0.55 is **outside** the 95% CI, and p-value is well below 0.05.

**We reject H₀.** The data suggests the true support level is significantly higher than the pollster's claimed 55%.

---
## 📌 Question 3 — Two-Proportion Z-Test (A/B Test Style)

| Group | n | Conversions |
|-------|---|-------------|
| Control | 5000 | 600 |
| Treatment | 5000 | 680 |

Test at **α = 0.01** whether the treatment **significantly improves** conversion rate.
Also compute the **99% CI** for the difference in proportions.

**Key concept to address:** When to use **pooled** vs **unpooled** SE?

---

#### Step 1 — Hypothesis

- **H₀:** p_ctrl = p_trt (treatment has no effect)
- **H₁:** p_trt > p_ctrl (treatment improves conversion — one-tailed)

One-tailed because we specifically care whether treatment is **better**, not just different.

#### Step 2 — Pooled vs Unpooled SE — The Key Distinction

This trips people up, so worth being clear:

| Situation | SE to use | Why |
|-----------|-----------|-----|
| **Hypothesis test** | Pooled | Under H₀ we assume p_ctrl = p_trt, so we estimate a single common proportion |
| **Confidence Interval** | Unpooled | We're estimating the actual difference — H₀ is no longer assumed, each group gets its own p̂ |

In [13]:
# Q3 given values
nc, xc = 5000, 600   # control: sample size, conversions
nt, xt = 5000, 680   # treatment: sample size, conversions
a3 = 0.01        # significance level

In [15]:
# individual sample proportions
pc = xc / nc   # control proportion
pt = xt / nt   # treatment proportion
pc, pt

(0.12, 0.136)

#### Step 3 — Pooled Proportion & SE (for hypothesis test)

$$\hat{p}_{pool} = \frac{x_c + x_t}{n_c + n_t}$$

In [16]:
# pooled estimate — assumes H0: pc = pt
pp = (xc + xt) / (nc + nt)
pp

0.128

In [21]:
# pooled SE for the test statistic
se_pool = (pp * (1 - pp) * (1/nc + 1/nt)) ** 0.5
se_pool

0.006681796165702753

In [23]:
z_q3 = (pt - pc) / se_pool
z_q3

2.3945657130528804

#### Step 4 — Critical Value (one-tailed, α = 0.01)

One-tailed → rejection region is only on the right side.
We need z such that P(Z < z) = 0.99

In [30]:
# one-tailed critical value at alpha=0.01
zcrit_q3_test = norm.ppf(1 - a3)
zcrit_q3_test

np.float64(2.3263478740408408)

**Decision:** z = 2.39 > z_critical = 2.33 → **Reject H₀**

#### Step 5 — p-value (one-tailed)

In [29]:
# one-tailed p-value: P(Z > z)
pval_q3 = norm.sf(z_q3)
pval_q3

np.float64(0.008320030697044572)

**Decision:** p = 0.0083 < α = 0.01 → **Reject H₀** ✓

#### Step 6 — 99% Confidence Interval (Unpooled SE)

$$CI = (\hat{p}_t - \hat{p}_c) \pm z_{\alpha/2} \cdot \sqrt{\frac{\hat{p}_c(1-\hat{p}_c)}{n_c} + \frac{\hat{p}_t(1-\hat{p}_t)}{n_t}}$$

Here we use **unpooled SE** because we're estimating the actual difference between two proportions — no assumption of equality.

In [31]:
# unpooled SE — each group uses its own proportion
se_unpool = np.sqrt(pc*(1-pc)/nc + pt*(1-pt)/nt)
se_unpool

np.float64(0.006679880238447394)

In [36]:
# two-tailed critical value for 99% CI
zcrit_q3_ci = norm.ppf(1 - a3/2)
zcrit_q3_ci

np.float64(2.5758293035489004)

In [38]:
diff_q3 = pt - pc   # observed difference in proportions
lo_q3   = diff_q3 - zcrit_q3_ci * se_unpool
hi_q3   = diff_q3 + zcrit_q3_ci * se_unpool
(lo_q3, hi_q3)

(np.float64(-0.0012062312623899972), np.float64(0.03320623126239003))

#### ✅ Final Conclusion — Q3

| Check | Result |
|-------|--------|
| Control p̂ | 0.12 (12%) |
| Treatment p̂ | 0.136 (13.6%) |
| z-stat (pooled SE) | 2.39 |
| Critical value (one-tailed) | 2.33 |
| p-value | 0.0083 |
| 99% CI (unpooled SE) | (-0.0012, 0.0332) |

**We reject H₀.** The treatment group shows a statistically significant improvement in conversion rate at α = 0.01.

**However**, notice that the 99% CI barely includes 0 — the lower bound is slightly negative (-0.0012). This is because the CI uses a two-tailed z (2.576) while the test was one-tailed (2.326). The two methods are answering slightly different questions here — the hypothesis test was directional, the CI is not. Both are valid, but comparing them directly can be misleading.

---
### Pooled vs Unpooled — Quick Summary

- **Pooled SE** → for the **test statistic** only, because under H₀ we assume the two groups have the *same* underlying proportion. Merging them gives a better estimate of that common p.
- **Unpooled SE** → for **confidence intervals**, because we're no longer assuming equality — we want to estimate how far apart the two proportions actually are.